geode-core/src/main/java/org/apache/geode/internal/cache/LocalRegion.java

```diff
@@ -5737,21 +5737,27 @@ public void checkIfAboveThreshold(final EntryEventImpl entryEvent) throws LowMem
    * @throws LowMemoryException if the target member for this operation is sick
    */
   private void checkIfAboveThreshold(final Object key) throws LowMemoryException {
-    if (this.memoryThresholdReached.get()) {
-      Set<DistributedMember> membersThatReachedThreshold = getMemoryThresholdReachedMembers();
-
+    MemoryThresholdInfo info = getAtomicThresholdInfo();
+    boolean thresholdReached = info.isMemoryThresholdReached();
+    Set<DistributedMember> membersThatReachedThreshold = info.getMembersThatReachedThreshold();
+    if (thresholdReached) {
       // #45603: trigger a background eviction since we're above the the critical
       // threshold
       InternalResourceManager.getInternalResourceManager(this.cache).getHeapMonitor()
           .updateStateAndSendEvent();
-
       throw new LowMemoryException(
           LocalizedStrings.ResourceManager_LOW_MEMORY_IN_0_FOR_PUT_1_MEMBER_2.toLocalizedString(
               getFullPath(), key, membersThatReachedThreshold),
           membersThatReachedThreshold);
     }
   }
 
+  protected MemoryThresholdInfo getAtomicThresholdInfo() {
+    MemoryThresholdInfo info = new MemoryThresholdInfo(this.memoryThresholdReached.get(),
```

Comment:
In your description you make it sound like LocalRegion doing this without synchronization is a problem. But it looks like your fix only does the sync on DistributionRegion. Also a copy of the HashSet is made on DistributedRegion but on LocalRegion a reference to the CopyOnWriteHashSet is used. I just wanted to check if it was okay for LocalRegion to be inconsistent. Couldn't multiple threads on a LocalRegion end up having concurrency issues?

Reply:
Yes, LocalRegion doing this without synchronization was the problem because DistributedRegion was inheriting and running that unsynchronized code.

LocalRegion does not keep a set like DistributedRegion does. When running with an instance of local region rather than distributed region, when it wants to throw a LowMemoryException and needs a set of members that are low it creates a new singleton collection with itself. klund and I looked at this and thought the code in LocalRegion was ok.

DistributedRegion has methods that are called by the NotificationHandler thread that can change the boolean (from LocalRegion) and the set (from DistributedRegion), but it does so with synchronization. This is why it needed a new copy of the set, so it couldn't be modified after getting a "snapshot" of a consistent view of the boolean/set pair.